### DISCUSSION TITLE
Minimizing Windows.

### DISCUSSION THREAD

**Discussion by Mauitron** [hyprwm/Hyprland#8281] [Show and tell]
i honestly don't know if this is anything to share. I have created a minimize-manager that allow you to (on my setup) press meta + m to minimize any active window and you could then press meta + I to open up the menu. the minimized windows will open on the workplace you are on, not the one they were started on.

oh, you could also click the icon to bring up the manager if you like using the mouse more than shortcuts. 

anyway. here is what it looks like. it is using wofi for the menu and search. 
![output_optimized](https://github.com/user-attachments/assets/57096f15-d5af-42a4-9c18-fc0e171a2287)

**Comment by michaeltraxler** [comment]
I would love to have such a feature. Where can I find the code for that?

**Reply by Mauitron** [reply]
Hi, i will get right on it! I just need to extract it from my nixos config. you could find it there yourself as i have a public repo of a generalized, somwhat stripped down version of my config. but that might be too much.

would you prefer it being rofi-wayland or wofi? I am in the midst of rewriting a lot of my scripts to either use rofi-wayland or just the rust std and eww for ui. So, i can more or less give you one of the three, depending on want and time you would be willing to wait.

**Reply by Mauitron** [reply]
also, do you want the waybar toggle aswell? or are you fine with just have the script to minimize and return the windows?

**Reply by Mauitron** [reply]
i have extracted the code and made a public repo. feel free to try it out! and don't hesitate to contact me if there is anything missing or anything you want added back. 

https://github.com/Mauitron/NiflVeil

**Reply by Arbifoxx** [reply]
is it possible to use this with rofi-wayland/wofi for the ui rather than eww?

**Comment by michaeltraxler** [comment]
This is so nice of you, to make this public.
It works like a charm and does exactly what I need!
This makes my day! Hyprland is great, but the "minimize" ("hide") feature is really missing. This functionality should be included in the standard code! 
One more small feature request:
For very fast short hiding and recover of windows (without resizing them, which would be annoying!) it would be great to have one more option in the script. An options like "restore_last", so skipping the wofi requester, just bring back the last window in the list back to the current screen. Then I can have an extra keyboard shortcut for that.
Save one short distraction and save one keypress to quickly "hide" and "unhide" a windows you are working in.

**Reply by Mauitron** [reply]
Oh, you are very welcome. I am very glad you like it! 
I agree that an easy way to minimize is really nice at times, the mental model of a workspace is not really the same as having a drawer to put things in.  

It worked? Great! I was afraid i might have missed extracting something from my configuration.nix. 

That is actually a really good idea. putting the order in an array you could use the index to call whatever hidden window you would want, as well as having last minimized and first open functionality. i will add it in! with some bells and whistles. although, I might as well start rewriting it in Rust while i am at it. if you have any ideas please feel free to write one or more issues on the repo, this way i won't forget over the holidays.

**Reply by Mauitron** [reply]
Just to keep you updated. i have finished converting the script into a rust program. it feels a lot nicer i think. i will start work on the UI and adding your suggested quick restore function. 

I will be using eww only for rendering the UI, all functionality will be in rust. this should result in a window minimizer that is not only more light weight than wofi, but also easier to customize, as you could just change the scss or yuck file. sounds good?

**Reply by michaeltraxler** [reply]
This sounds very good! I'm using your shell-based minimizing now all the time. For the rust-based version: Did you publish it? Can I try it out?

**Reply by Mauitron** [reply]
I am very happy to hear it! Hopefully you'll like the new one even more. Are you able to build with cargo? 

I was having a bit of problem with the UI as I am new to EWW, but It is going great. I have been a bit slow since Christmas as i am not home, i am in the other side of the country  visiting my family. I will be going back home on Tuesday so I will be mush more active then, it is quite cumbersome programming on a slow laptop. but here is a picture of the interface so far, you can see it doesn't look that great at the moment. 
  
![image](https://github.com/user-attachments/assets/bc6ce515-63cd-4b71-a53e-5ec3248cdf32)

**Reply by Mauitron** [reply]
i have added your suggested functionality. it can now bring back the windows without going through the menu. the shortcut opens them in the order they were minimized, meaning, last mimimized, first out.

i will now work on the ui.

**Reply by michaeltraxler** [reply]
That is great! Thanks a lot for your work!
I guess also others will appreciate this!
And yes, I can build rust applications with cargo.

**Reply by Mauitron** [reply]
no worries! I like the work ^^

i will finish up the code and updating the GitHub page over the weekend. should be done by tomorrow or Sunday at the latest, unless attacked by gremlins of-course.

**Reply by Mauitron** [reply]
i have uploaded the rust version and i have edited the github page. gremlins did attack and now cargo isn't able to parse the cargo.toml for me, my cursor is also gone for some reason. The code should compile and work though, if it doesn't please let me know. and please give some feedback once you have tried it out.

**Comment by nazya** [comment]
My solution is to send  a window to +10 workspace by

`bind = $mainMod, N, movetoworkspacesilent, +10`

and then restore by using fzf

`bind = $mainMod, M, exec, [stayfocused;float;size 1700 900] foot --font='CodeNewRoman Nerd Font Mono:size=16' -e bash -c 'active=$(hyprctl activeworkspace -j | jq ".id"); address=$(hyprctl -j clients | jq -r --argjson target_ws $((10 + active)) ".[] | select(.workspace.id == \$target_ws) | \"\(.address)\t\(.title)\"" | fzf --preview "addr={1}; hyprctl -j clients | jq -r \".[] | select(.address == \\\"\$addr\\\") | \\\"address: \(.address)\ntitle: \(.title)\nclass: \(.class)\nworkspace: \(.workspace.id)\ntags: \(.tags)\\\"\"" --preview-window=down:50%:wrap --layout=reverse -d $"\t" --with-nth=2 | awk -F"\t" "{print \$1}"); hyprctl dispatch movetoworkspacesilent $active,address:$address; hyprctl dispatch focuswindow address:$address'`

For the sake of simplicity one can script the following commands using fuzzel instead of fzf

```
active=$(hyprctl activeworkspace -j | jq '.id')

address=$(hyprctl -j clients | jq -r --argjson target_ws $((10 + $(hyprctl activeworkspace -j | jq '.id'))) '.[] | select(.workspace.id == $target_ws) | "\(.address)\t\(.title)"' | fuzzel --dmenu | awk -F'\t' '{print $1}')

hyprctl dispatch movetoworkspacesilent $active,address:$address
hyprctl dispatch focuswindow address:$address
```

but it seems that --with-nth=2 feature has buggy implementation in fuzzel.

**Reply by Mauitron** [reply]
That is cleaver approach approach, if you are interested in a more custom built approach,  i send them to secret work-spaces, this way they you won't scroll past them and whatnot, you can implement that yourself in your script if you want. otherwise feel free no try out Nifelveil. https://github.com/Mauitron/NiflVeil

**Reply by nazya** [reply]
gonna check the secret workspaces, great idea!

btw, i've just found a way to preview windows -- fzf  preview command is used to restore a window in a floating mode and the fzf "change" bind is used to send it back on selection change in fzf. works great for me (master layout)

`bind = $mainMod, M, exec, [stayfocused;bordersize 0;float;size 30% 100%;move 70% 0%] foot --font='CodeNewRoman Nerd Font Mono:size=16' -e bash -c 'active=$(hyprctl activeworkspace -j | jq ".id"); tmpfile="/dev/shm/fzf_prev_addr"; > "$tmpfile"; address=$(hyprctl -j clients | jq -r --argjson target_ws $((10 + active)) ".[] | select(.workspace.id == \$target_ws) | \"\(.address)\t\(.title)\t\(.workspace.id)\"" | fzf --wrap --delimiter=$"\t" --with-nth=2 --preview "addr={1}; ws={3}; if [ -f \"$tmpfile\" ]; then prev_addr=\$(cat $tmpfile); if [ -n \"\$prev_addr\" ] && [ \"\$prev_addr\" != \"\$addr\" ]; then hyprctl dispatch movetoworkspacesilent $((10 + active)),address:\$prev_addr > /dev/null; fi; fi; hyprctl dispatch setfloating address:\$addr > /dev/null; hyprctl dispatch movewindowpixel exact 0% 0%, address:\$addr > /dev/null; hyprctl dispatch resizewindowpixel exact 70% 100%, address:\$addr > /dev/null; hyprctl dispatch movetoworkspacesilent $active,address:\$addr > /dev/null; echo \$addr > $tmpfile; hyprctl -j clients | jq -r \".[] | select(.address == \\\"\$addr\\\") | \\\"address: \(.address)\ntitle: \(.title)\nclass: \(.class)\nworkspace: \(.workspace.id)\ntags: \(.tags)\\\"\"" --bind "ctrl-c:execute-silent( if [ -f $tmpfile ]; then addr=\$(cat $tmpfile); hyprctl dispatch movetoworkspacesilent $((10 + active)),address:\$addr; fi; )+abort" --preview-window=down:30%:wrap --layout=reverse | awk -F"\t" "{print \$1}"); hyprctl dispatch togglefloating address:$address; hyprctl dispatch focuswindow address:$address; rm -f \"$tmpfile\"'`

**Reply by Mauitron** [reply]
No problem, I think that if you hide the windows in the secret workplace, it would feel more invisible and 'native'. although, i am ofcourse biased, as it was the approach I took. 

That is a great addition! i would recommend you, if you aren't doing so already, to put the command into a .sh and bind that to the key instead. This way, it would be a less intimidating bind and it would have a clearer way of reading/modifying it for other users.

Keep up the great work!

**Comment by Simon-Martens** [comment]
I had a similar problem: I needed to minimize windows, which I don't use (except for corporate purposes, like microsoft teams), and needed these windows to stay out of my sight, but still receive notifications. Here's a small application that integrates with waybars tray module: https://github.com/Simon-Martens/hyprland-minimizer It works by moving the window to a special workspace "minimized" and nofifies waybar over the DBus about the minimized application. If someone decides to write something more proper, or the functionality (with integrating into org.kde.StatusNotifierWatcher) somehow finds it's way into wayland, that would be nice! Until then...

**Comment by SykikXO** [comment]
hello people, after multiple updates to hyprland (gestures) found a new method to implement the **3 finger minimize** from windows on hyprland. just add these two files to your hyprland config and ignore the stuff written above

`gesture = 3, down, dispatcher, movetoworkspacesilent, special:minimized`
`gesture = 3, up, dispatcher, exec, hyprctl --batch "dispatch togglespecialworkspace minimized; dispatch movetoworkspace +0"`



i know of a annoying bug of doing thing this way, if you accidently 3finger swipe up, it opens a special workspace that "dims" screen,
so here's a better version


'gesture = 3, down, dispatcher, movetoworkspacesilent, special:minimized'

^ looks for 3 finger down gesture, if found puts current selected window to a special workspace called minimized

'gesture = 3, up, dispatcher, exec, hyprctl workspaces -j | jq -e '.[] | select(.name=="special:minimized")' >/dev/null && hyprctl --batch "dispatch togglespecialworkspace minimized; dispatch movetoworkspace +0"'

^ looks for 3 finger up gesture, if found, looks for a special workspace called minimized (if it doesnt have any windows in it, it wont exist) and retrieves the last pushed window to current workspace (stack??) 



/////
**UPDATE v0.55 LUA CONFIG**
/////




this thingy for minimize with 3 finger down,

`
hl.gesture({
	fingers = 3,
	direction = "down",
	action = function()
		hl.dispatch(hl.dsp.window.move({ workspace = "special:minimized", follow = false }))
	end,
})`


this thingy for un-minimize with 3 finger up.

`
hl.gesture({
	fingers = 3,
	direction = "up",
	action = function()
		hl.dispatch(
			hl.dsp.exec_cmd(
				[[hyprctl workspaces -j | jq -e '.[] | select(.name=="special:minimized")' >/dev/null && hyprctl dispatch 'hl.dsp.workspace.toggle_special("minimized")' && hyprctl dispatch 'hl.dsp.window.move({ workspace = "+0" })']]
			)
		)
	end,
})`



or you can just use the "action = function() asdfasdfjasldfhjashdflk" thingy with a keybind so you can minimize and unminimize with a keybind instead of gesture like below

for minimize 

`
hl.bind(mainMod .. "+ Z", hl.dsp.window.move({ workspace = "special:minimized", follow = false }))`

for maximize 

hl.bind(mainMod .. " + SHIFT + Z",function()
  local ws = hl.get_workspace("special:minimized")
  if ws and ws.windows > 0 then
    hl.dispatch(hl.dsp.workspace.toggle_special("minimized"))
    hl.dispatch(hl.dsp.window.move({ workspace = "+0" }))
  end
end)

//functionality exactly same as the older implementation above

**Comment by rollecode** [comment]
If you use [DankMaterialShell](https://github.com/AvengeMedia/DankMaterialShell) with Hyprland and [hyrpbars](https://code.hyprland.org/hyprwm/hyprland-plugins/src/branch/hyprbars/hyprbars), there's a way with [dms-minimize](https://github.com/ronilaukkarinen/dms-minimize).
